### PR TITLE
SERVER-78199: btree index check, avoid create duplicate btree indexes, Avoid affecting performance

### PR DESCRIPTION
**Pull request by y123456yz**
btree index check, avoid create duplicate btree indexes, Avoid affecting performance

btree index check, avoid create duplicate btree indexes, Avoid affecting performance by the deplicate index.
for example:
 add two index: db.collection.createIndex({a:1}) and db.collection.createIndex({a:11})
 the tow index are actually the same, One of them is a useless index, but it can affect the insert performance.

 In addition, They also affecting query performance because both are candidate indexes. This increases storage consts.

### CODE REVIEW

**Comment by y123456yz** [comment]
https://jira.mongodb.org/browse/SERVER-78199

File: src/mongo/db/index/index_descriptor.cpp [hunk 1]
```diff
@@ -171,8 +171,36 @@ IndexDescriptor::Comparison IndexDescriptor::compareIndexOptions(
     const IndexCatalogEntry* existingIndex) const {
     auto existingIndexDesc = existingIndex->descriptor();
 
+    //btree index check, avoid create duplicate btree indexes, Avoid affecting performance by the deplicate index.
+    //for example:
+    // add two index: db.collection.createIndex({a:1}) and db.collection.createIndex({a:11})
+    // the tow index are actually the same, One of them is a useless index, but it can affect the insert performance.
+    //
+    // In addition, They also affecting query performance because both are candidate indexes.
+    auto dealBtreeKeyPattern = [&](const BSONObj& indexKeyPattern) {
```
Comment (debdutdeb) [inline src/mongo/db/index/index_descriptor.cpp:180]: why is this a lambda? 🤔

**Review by mongodt — COMMENTED** [review]
"btree" shouldn't be mentioned in both comments and code (variable name or function names).

File: src/mongo/db/index/index_descriptor.cpp [hunk 2]
```diff
@@ -171,8 +171,41 @@ IndexDescriptor::Comparison IndexDescriptor::compareIndexOptions(
     const IndexCatalogEntry* existingIndex) const {
     auto existingIndexDesc = existingIndex->descriptor();
 
+    //btree index check, avoid create duplicate btree indexes, Avoid affecting performance by the deplicate index.
```
Comment (mongodt) [inline src/mongo/db/index/index_descriptor.cpp:174]: avoid create -> avoid creating.

"Avoid affecting performance by the deplicate index." is not necessary as it's obvious and performance is not the only side effect.

File: src/mongo/db/index/index_descriptor.cpp [hunk 3]
```diff
@@ -171,8 +171,41 @@ IndexDescriptor::Comparison IndexDescriptor::compareIndexOptions(
...
+    //for example:
+    // add two index: db.collection.createIndex({a:1}) and db.collection.createIndex({a:11})
```
Comment (mongodt) [inline src/mongo/db/index/index_descriptor.cpp:176]: two index->two indexes

File: src/mongo/db/index/index_descriptor.cpp [hunk 4]
```diff
@@ -171,8 +171,41 @@ IndexDescriptor::Comparison IndexDescriptor::compareIndexOptions(
...
+    // the tow index are actually the same, One of them is a useless index, but it can affect the insert performance.
+    //
+    // In addition, They also affecting query performance because both are candidate indexes.
```
Comment (mongodt) [inline src/mongo/db/index/index_descriptor.cpp:179]: remove

File: src/mongo/db/index/index_descriptor.cpp [hunk 5]
```diff
@@ -171,8 +171,41 @@ IndexDescriptor::Comparison IndexDescriptor::compareIndexOptions(
...
+        }
+        
+        return build.obj();
+    };
+    
+    BSONObj newKeyPattern = dealBtreeKeyPattern(keyPattern());
+
     // We first check whether the key pattern is identical for both indexes.
-    if (SimpleBSONObjComparator::kInstance.evaluate(keyPattern() !=
+    if (SimpleBSONObjComparator::kInstance.evaluate(newKeyPattern =
```
Comment (mongodt) [inline src/mongo/db/index/index_descriptor.cpp:208]: = or == ?

**Comment by y123456yz** [comment]
@mongodt 
thanks for you guidance, I have made the changes according to your guidance.
please review again.

**Comment by y123456yz** [comment]
@mongodt 
master branch PR:
[SERVER-84667](https://jira.mongodb.org/browse/SERVER-84667)

**Pull request closed without merging**